File: _739_Daily_Temperatures.java
import java.util.Arrays;

class Solution {
    public int[] dailyTemperatures(int[] T) {
        int len = T.length;
        int[] ans = new int[len];
        int right_max = Integer.MIN_VALUE;
        int tmp;
        for (int i=len-1; i >= 0; --i) {
            if (right_max <= T[i]) {
                right_max = T[i];
            }
            else {
                tmp = 1;
                while (T[i+tmp] <= T[i]) {
                    tmp += ans[i+tmp];
                }
                ans[i] = tmp;
            }
        }
        return ans;
    }
}

public class _739_Daily_Temperatures {
    public static void main(String[] args) {
        Solution s = new Solution();
        int[] t = {73, 74, 75, 71, 69, 72, 76, 73};
        System.out.println(Arrays.toString(s.dailyTemperatures(t)));
    }
}

File: 739. Daily Temperatures.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    vector<int> dailyTemperatures(vector<int>& T) {
        vector<int> ans(T.size(), 0);
        if (T.size() == 0) return ans;
        ans[T.size()-1] = 0;
        int right_max = T[T.size()-1];
        int tmp;
        for (int i=T.size()-2; i >= 0; --i) {
            if (T[i] >= right_max) {
                right_max = T[i];
            }
            else {
                tmp = 1;
                while (T[i+tmp] <= T[i]) {
                    tmp += ans[i+tmp];
                }
                ans[i] = tmp;
            }
        }
        return ans;
    }
};

int main() {
    vector<int> v{1,2,3,4,5,6,7,6,5,4,3,2,1};
    auto ans = Solution().dailyTemperatures(v);
    for (auto i: ans) cout << i << ' ';
    cout << endl;
    return 0;
}
